Mathlib/Order/Lattice.lean

```diff
@@ -0,0 +1,1469 @@
+    _ = x ⊓ (x ⊓ y ⊔ z) := by simp only [inf_assoc, sup_inf_right, eq_self_iff_true]
+    _ = (x ⊔ x ⊓ y) ⊓ (x ⊓ y ⊔ z) := by rw [sup_inf_self]
+    _ = (x ⊓ y ⊔ x) ⊓ (x ⊓ y ⊔ z) := by rw [sup_comm]
+    _ = x ⊓ y ⊔ x ⊓ z := by rw [sup_inf_left]
+
+#align inf_sup_left inf_sup_left
+
+instance OrderDual.distribLattice (α : Type _) [DistribLattice α] : DistribLattice αᵒᵈ where
+  __ := inferInstanceAs (Lattice αᵒᵈ)
+  le_sup_inf := fun _ _ _ => le_of_eq (@inf_sup_left α).symm
+
+theorem inf_sup_right : (y ⊔ z) ⊓ x = y ⊓ x ⊔ z ⊓ x := by
+  simp only [inf_sup_left, fun y : α => @inf_comm α _ y x, eq_self_iff_true]
+#align inf_sup_right inf_sup_right
+
+theorem le_of_inf_le_sup_le (h₁ : x ⊓ z ≤ y ⊓ z) (h₂ : x ⊔ z ≤ y ⊔ z) : x ≤ y :=
+  calc
+    x ≤ y ⊓ z ⊔ x := le_sup_right
+    _ = (y ⊔ x) ⊓ (x ⊔ z) := by rw [sup_inf_right, @sup_comm _ _ x]
+    _ ≤ (y ⊔ x) ⊓ (y ⊔ z) := inf_le_inf_left _ h₂
+    _ = y ⊔ x ⊓ z := sup_inf_left.symm
+    _ ≤ y ⊔ y ⊓ z := sup_le_sup_left h₁ _
+    _ ≤ _ := sup_le (le_refl y) inf_le_left
+
+#align le_of_inf_le_sup_le le_of_inf_le_sup_le
+
+theorem eq_of_inf_eq_sup_eq {α : Type u} [DistribLattice α] {a b c : α} (h₁ : b ⊓ a = c ⊓ a)
+  (h₂ : b ⊔ a = c ⊔ a) :
+    b = c :=
+  le_antisymm (le_of_inf_le_sup_le (le_of_eq h₁) (le_of_eq h₂))
+    (le_of_inf_le_sup_le (le_of_eq h₁.symm) (le_of_eq h₂.symm))
+#align eq_of_inf_eq_sup_eq eq_of_inf_eq_sup_eq
+
+end DistribLattice
+
+-- See note [reducible non-instances]
+/-- Prove distributivity of an existing lattice from the dual distributive law. -/
+@[reducible]
+def DistribLattice.ofInfSupLe [Lattice α] (inf_sup_le : ∀ a b c : α, a ⊓ (b ⊔ c) ≤ a ⊓ b ⊔ a ⊓ c) :
+  DistribLattice α where
+  __ := inferInstanceAs (Lattice αᵒᵈ)
+  __ := @OrderDual.distribLattice αᵒᵈ {inferInstanceAs (Lattice αᵒᵈ) with le_sup_inf := inf_sup_le}
+#align distrib_lattice.of_inf_sup_le DistribLattice.ofInfSupLe
+
+/-!
+### Lattices derived from linear orders
+-/
+
+
+-- see Note [lower instance priority]
+instance (priority := 100) LinearOrder.toLattice {α : Type u} [o : LinearOrder α] : Lattice α :=
+  { o with
+    sup := max,
+    le_sup_left := le_max_left, le_sup_right := le_max_right, sup_le := fun _ _ _ => max_le,
+    inf := min,
+    inf_le_left := min_le_left, inf_le_right := min_le_right, le_inf := fun _ _ _ => le_min }
+#align linear_order.to_lattice LinearOrder.toLattice
+
+section LinearOrder
+
+variable [LinearOrder α] {a b c d : α}
+
+theorem sup_eq_max : a ⊔ b = max a b :=
+  rfl
+#align sup_eq_max sup_eq_max
+
+theorem inf_eq_min : a ⊓ b = min a b :=
+  rfl
+#align inf_eq_min inf_eq_min
+
+theorem sup_ind (a b : α) {p : α → Prop} (ha : p a) (hb : p b) : p (a ⊔ b) :=
+  (IsTotal.total a b).elim (fun h : a ≤ b => by rwa [sup_eq_right.2 h]) fun h => by
+  rwa [sup_eq_left.2 h]
+#align sup_ind sup_ind
+
+@[simp]
+theorem le_sup_iff : a ≤ b ⊔ c ↔ a ≤ b ∨ a ≤ c := by
+  -- Porting note: why is this necessary?
+  have := inferInstanceAs (IsTotal α (· ≤ ·))
```

Comment:
Also a bit worried here. Without this have statement it's looking for the exact same instances but can't find it.

Reply:
What if you use the Lean4 version of `le_total`?

Reply:
`le_total` works. Still strange that `total_of` doesn't, though